src/lib.rs

```diff
@@ -506,43 +506,71 @@ mod opassign {
 
     impl<T: Clone + Integer + NumAssign> AddAssign for Ratio<T> {
         fn add_assign(&mut self, other: Ratio<T>) {
-            self.numer *= other.denom.clone();
-            self.numer += self.denom.clone() * other.numer;
-            self.denom *= other.denom;
+            if self.denom == other.denom {
+                self.numer += other.numer
+            } else {
+                let lcm = self.denom.lcm(&other.denom.clone());
+                let lhs_numer = self.numer.clone() * (lcm.clone() / self.denom.clone());
+                let rhs_numer = other.numer * (lcm.clone() / other.denom);
+                self.numer = lhs_numer + rhs_numer;
+                self.denom = lcm;
+            }
             self.reduce();
         }
     }
 
+    // (a/b) / (c/d) = (a/gcd_ac)*(d/gcd_bd) / ((c/gcd_ac)*(b/gcd_bd))
     impl<T: Clone + Integer + NumAssign> DivAssign for Ratio<T> {
         fn div_assign(&mut self, other: Ratio<T>) {
-            self.numer *= other.denom;
-            self.denom *= other.numer;
-            self.reduce();
+            let gcd_ac = self.numer.gcd(&other.numer.clone());
+            let gcd_bd = self.denom.gcd(&other.denom.clone());
+            self.numer /= gcd_ac.clone();
+            self.numer *= other.denom / gcd_bd.clone();
+            self.denom /= gcd_bd.clone();
+            self.denom *= other.numer / gcd_ac.clone();
+            self.reduce(); //TODO: remove this line. see #8.
```

Comment:
Correct me if I'm wrong, but I think we can only remove that `reduce()` if the inputs were properly normalized. While I think we do want normalized output everywhere, it's less certain that we should assume that about inputs. But, we can debate that in #8, I suppose, before removing actually this.

Reply:
There is some "built-in" reduction because of greatest common denominators. For this particular function, any factors in common between `a` and `c` will be taken out (same for `b` and `d`). This means, for example, that `2/4 ÷ 10/6` -> `3/10` (rather than `12/40`). After this operation, a call to reduce would be wasteful. We do get some reduction from this algorithm for free, but it is imperfect. 

The only "perfect" way to deal with un-reduced `Ratio`s would be to `reduce()` the incoming `Ratio`s, because otherwise the operation might overflow when the result would not.

I think we're on the same page about this, but I'm going to spell it out just to be sure. After removing that call to `reduce()` we would be able to guarantee the output is reduced if the inputs were reduced. But even with that `reduce()` at the end, intermediate calculations in an operation could overflow where the result is in range. However, if the inputs are reduced, this algorithm guarantees that the outputs too will be reduced.